hazelcast/src/main/java/com/hazelcast/cp/event/impl/CPGroupAvailabilityEventGracefulImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cp.event.impl;
+
+import com.hazelcast.cp.CPGroup;
+import com.hazelcast.cp.CPGroupId;
+import com.hazelcast.cp.CPMember;
+import com.hazelcast.cp.event.CPGroupAvailabilityEvent;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+
+import static com.hazelcast.internal.serialization.impl.SerializationUtil.readCollection;
+import static com.hazelcast.internal.serialization.impl.SerializationUtil.writeCollection;
+
+// it's probably easier to just add a flag to the existing availability message, e.g. isShutdown -- default value of false
+public class CPGroupAvailabilityEventGracefulImpl implements CPGroupAvailabilityEvent, IdentifiedDataSerializable {
+    private CPGroupId groupId;
+    private Collection<CPMember> members;
+    private Collection<CPMember> unavailableMembers;
+
+    public CPGroupAvailabilityEventGracefulImpl() {
+    }
+
+    public CPGroupAvailabilityEventGracefulImpl(CPGroupId groupId, Collection<CPMember> members,
+                                                Collection<CPMember> unavailableMembers) {
+        this.groupId = groupId;
+        this.members = members;
+        this.unavailableMembers = unavailableMembers;
+    }
+
+    @Override
+    public CPGroupId getGroupId() {
+        return groupId;
+    }
+
+    @Override
+    public Collection<CPMember> getGroupMembers() {
+        return members;
+    }
+
+    @Override
+    public Collection<CPMember> getUnavailableMembers() {
+        return unavailableMembers;
+    }
+
+    @Override
+    public int getMajority() {
+        return members.size() / 2 + 1;
+    }
+
+    @Override
+    public boolean isMajorityAvailable() {
```

Comment:
We should treat the case with the 2 remaining CP members as a separate case, for example somehow like this:

```
 Set<CPMember> leftMembers = new HashSet<>(members);
leftMembers.removeAll(missingMembers);
return leftMembers.size() > 1 ? missingMembers.size() < getMajority() : false;
```

And it is also actually during the non-graceful shutdown for `CPGroupAvailabilityEventImpl`.

Let's consider the 3 CP members cluster with non-graceful shutdown:
- 3rd member has been shutdown -> the `availabilityDecreased` event is published
- after the `missing-cp-member-auto-removal-seconds` the 3rd member is auto-removed, the CP group shrink and the majority becomes equal 2, the `memberRemoved` event is published
- 2nd member shutdown -> NO `availabilityDecreased` event is published (since the 2nd member could not be removed from the CP group)

Also, suggest adding a test for this scenario